components/authz-service/storage/postgres/migration/legacy/iam.go

```diff
@@ -1,45 +1,73 @@
-	storage_v1 "github.com/chef/automate/components/authz-service/storage/v1"
-	storage "github.com/chef/automate/components/authz-service/storage/v2"
+	constants_v1 "github.com/chef/automate/components/authz-service/storage/postgres/migration/legacy/constants/v1"
+	constants_v2 "github.com/chef/automate/components/authz-service/storage/postgres/migration/legacy/constants/v2"
+	"github.com/chef/automate/lib/logger"
 	uuid "github.com/chef/automate/lib/uuid4"
+	"github.com/pkg/errors"
 )
 
-// Note: it would be silly to translate the migration status into errors twice,
-// so this method is an exception to the "only grpc handlers return status
-// instances" rule.
-func (s *policyServer) okToMigrate(ctx context.Context, ms storage.MigrationStatus) error {
-	switch ms {
-	case storage.Successful, storage.SuccessfulBeta1:
-		return status.Error(codes.AlreadyExists, "already migrated")
-	case storage.InProgress:
-		return status.Error(codes.FailedPrecondition, "migration already in progress")
-	case storage.Failed:
-		s.log.Info("migration failed before; attempting again")
+// MigrateToV2 inserts needed IAM v2 resources into the db and
+// migrates any valid v1 policies
+func MigrateToV2(ctx context.Context, db *sql.DB) error {
+	l, err := logger.NewLogger("text", "info")
+	if err != nil {
+		return errors.Wrap(err, "could not initialize logger")
+	}
+	for _, role := range defaultRoles() {
+		if err := createRole(ctx, db, &role); err != nil {
+			return errors.Wrapf(err,
+				"could not create default role with ID: %s", role.ID)
+		}
+	}
+
+	for _, pol := range v2DefaultPolicies() {
+		if _, err := createV2Policy(ctx, db, &pol); err != nil {
+			return errors.Wrapf(err,
+				"could not create default policy with ID: %s", pol.ID)
+		}
+	}
+
+	errs, err := migrateV1Policies(ctx, db)
+	if err != nil {
+		return errors.Wrapf(err, "migrate v1 policies")
+	}
+
+	reports := []string{}
+	for _, e := range errs {
+		reports = append(reports, e.Error())
+	}
+	if len(reports) != 0 {
+		l.Infof("invalid v1 policies could not be migrated: %v", reports)
 	}
+
 	return nil
 }
 
+/*
+COPY PASTA DATABASE CODE
+
+The below is code we've copied from our database functionality because we need
+versions of the database functions needed for the migrations that do not change.
+This is because this migration is run at a single point in time as part of the schema
+upgrades. So this code need to be compatible with a specific schema version that never changes.
+*/
+
 // migrateV1Policies has two error returns: the second one is the ordinary,
 // garden-variety, "something went wrong, I've given up" signal; the first one
 // serves as an aggregate of errors that happened attempting to convert and
 // store individual (custom) policies.
-func (s *policyServer) migrateV1Policies(ctx context.Context) ([]error, error) {
-	pols, err := s.v1.ListPoliciesWithSubjects(ctx)
+func migrateV1Policies(ctx context.Context, db *sql.DB) ([]error, error) {
```

Comment:
Because this code can never be modified and we explicitly don't want people to extend or use this code. Leaving the interface would imply that this code is extensible, which it cannot be as it is migration code against a specific version of the schema.